### DISCUSSION TITLE
Using dnsdist as a stub resolver

### DISCUSSION THREAD

**Discussion by karelbilek** [PowerDNS/pdns#14811] [Q&A] [answered]
Hello.

I am using dnsmasq as a local stub resolver, but I start hitting some limits with what dnsmasq can do. (It's an amazing tool btw, but it has its limits because of what it is.)

I realized I can be using dnsdist a stub resolver pretty effectively. It has a cache; I can use other recursors as "backends"; I can even force use of DNS-over-HTTPS easily.

I know that dnsdist is intended as a load balancer, but... I don't see any downside of using it as a local stub resolver? It seems like a more robust solution than using something like stubby.

I don't see anyone on the internet writing about this usecase so maybe I am missing something

**Comment by karelbilek** [comment]
It seems folks at Cloudflare had exactly the same idea and it seems to be working for them!

https://indico.dns-oarc.net/event/31/contributions/687/attachments/663/1106/Ultimate_Stub-resolver.pdf

![Screenshot 2024-11-01 at 9 52 00](https://github.com/user-attachments/assets/6b64f818-d34f-4d05-aed5-90ed48c8e365)

Nice

**Reply by karelbilek** [reply]
Reading it more closely, it seems like they use it for some more advanced usage. But at least I am not the only one to think of dnsdist as a local stub resolver

**Comment by omoerbeek** [comment, marked as answer]
Maybe people are not posting about this because it *is* kinda obvious?

At least for me using dnsdist as a local caching stub resolver seems a pretty obvious use-case. You can throw in some extra resolver backends for added redundancy.

**Reply by karelbilek** [reply]
One thing that I have hit now is that it's impossible to set up servfail on timeouts with dnsdist

https://github.com/PowerDNS/pdns/issues/13575

**Reply by karelbilek** [reply]
It's possible to write retry rules

https://dnsdist.org/reference/dq.html#DNSResponse:restart

but those are NOT hit with timeouts.
